gateway-sp-comms/src/ereport.rs

```diff
@@ -0,0 +1,752 @@
+    packet: &[u8],
+) -> Result<(RestartId, Response), DecodeError> {
+    let (header, rest) =
+        gateway_messages::deserialize::<EreportResponseHeader>(packet)
+            .map_err(DecodeError::Header)?;
+    match header {
+        // Packet is empty
+        EreportResponseHeader::V0(ResponseHeaderV0 {
+            kind: ResponseKindV0::Empty,
+            restart_id,
+            ..
+        }) => Ok((restart_id, Response::Ereports(Vec::new()))),
+        // Packet is data
+        EreportResponseHeader::V0(ResponseHeaderV0 {
+            kind: ResponseKindV0::Data,
+            restart_id,
+            ..
+        }) => {
+            // V0 ereport packets consist of:
+            //
+            // - the first ENA in the packet
+            // - a CBOR list (using the "indeterminate") encoding, where each
+            //   entry is a CBOR list of 4 elements:
+            //      1. the name of the task that produced the ereport
+            //      2. the task's generation number
+            //      3. the system uptime in milliseconds
+            //      4. a CBOR object containing the rest of the ereport
+            //
+            // See RFD 545 4.4 for details:
+            // https://rfd.shared.oxide.computer/rfd/0545#_readresponse
+            let (start_ena, rest) = gateway_messages::deserialize::<Ena>(rest)
+                .map_err(DecodeError::Ena)?;
+            let cbor_ereports =
+                serde_cbor::from_slice::<Vec<Vec<CborValue>>>(rest)
+                    .map_err(DecodeError::EreportsDeserialize)?;
+
+            let mut json_ereports = Vec::with_capacity(cbor_ereports.len());
+            let mut task_names = Vec::new();
+            let mut ena = start_ena;
+            for (n, mut parts) in cbor_ereports.into_iter().enumerate() {
+                let ereport =
+                    parts.pop().ok_or(DecodeError::MalformedEreport {
+                        n,
+                        msg: "ereport list empty",
+                    })?;
+                let uptime =
+                    parts.pop().ok_or(DecodeError::MalformedEreport {
+                        n,
+                        msg: "missing Hubris uptime list entry",
+                    })?;
+                let task_gen =
+                    parts.pop().ok_or(DecodeError::MalformedEreport {
+                        n,
+                        msg: "missing task generation list entry",
+                    })?;
+                let task_name = match parts.pop() {
+                    Some(CborValue::Text(name)) => {
+                        task_names.push(name.clone());
+                        name
+                    }
+                    Some(CborValue::Integer(i)) => task_names
+                        .get(i as usize)
+                        .cloned()
+                        .ok_or(DecodeError::BadTaskNameIndex {
+                            n,
+                            index: i as usize,
+                        })?,
```

Comment:
@cbiffle this is what I was asking about in https://github.com/oxidecomputer/rfd/pull/849#discussion_r2027376877: right now, this code assumes that when an ereport's task name is an integer, it will always be the index of an ereport that was _earlier in the packet_ than the current one. I wanted to get your confirmation of whether we could rely on that assumption or would need to handle indexes pointing _ahead_ of the current ereport.